Refactor Save search modal

chrtorres opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##
- Refactored Save Search modal from legacy dependency libraries to Material UI.
- Applied form validation to prevent empty strings or duplicate entries when entering a new saved search.
- Updated the 'Update Saved Search' functionality.
- Removed 'vulnerabilityTemplate' and 'createVulnerability' from saved-search. Functionality no longer needed.

<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##
Prior to this fix, save search was reliant on legacy dependency libraries. In an effort to make the application uniform and simplified, this modal has been refactored using Material UI components. Also, When a user is able to save a search, there should be some sort of check to prevent the entry of duplicate names and/or empty names. With this fix, form validation is now present to avoid invalid entries.

This resolves [CRASM-629](https://maestro.dhs.gov/jira/browse/CRASM-629)


<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such -->
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##

<!-- How did you test your changes? How could someone else test this PR? -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->


## 📷 Screenshots  ##

#### Save Search Modal Before ####
<img width="528" alt="Screenshot 2024-09-11 at 10 45 40 AM" src="https://github.com/user-attachments/assets/a7c58041-97e1-4b80-831c-5d39c43da07a">

#### Save Search Modal Refactored ####
<img width="455" alt="Screenshot 2024-09-11 at 10 46 16 AM" src="https://github.com/user-attachments/assets/97d94cff-188a-4370-bc5d-cad84071f257">

#### Inventory page ####
<img width="1520" alt="Screenshot 2024-09-09 at 3 04 52 PM" src="https://github.com/user-attachments/assets/26194d82-b661-4429-a87b-5452d9cb4508">

#### Saving a new search ####

- Apply any filter to enable the "Save Search" button.
<img width="1520" alt="Screenshot 2024-09-09 at 3 37 47 PM" src="https://github.com/user-attachments/assets/f28931fc-eafc-4484-afab-28fb878d173c">

#### Applying saved search ####
<img width="1520" alt="Screenshot 2024-09-09 at 3 06 59 PM" src="https://github.com/user-attachments/assets/bcb8fc62-93f3-49a8-a8a4-412d9d95331f">

#### Updating a saved search ####

- When updating a saved search, select the search you would like to update, apply or remove desired filters, click "Update Saved Search".
<img width="1694" alt="Screenshot 2024-09-11 at 10 47 26 AM" src="https://github.com/user-attachments/assets/1f7ab927-9b34-4857-a227-302d9d606ad8">

#### Form Validation against empty strings ####
<img width="1520" alt="Screenshot 2024-09-09 at 3 39 17 PM" src="https://github.com/user-attachments/assets/fe3bef10-a678-4b46-8ae9-24bc9d2285f4">

#### Form validation against duplicate entries ####
<img width="1520" alt="Screenshot 2024-09-09 at 3 40 13 PM" src="https://github.com/user-attachments/assets/60412521-08d2-48f0-9bfe-2511af030bf4">



## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] *All* future TODOs are captured in issues, which are referenced
      in code comments.
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [ ] Tests have been added and/or modified to cover the changes in this PR.
- [ ] All new and existing tests pass.

## ✅ Pre-merge checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- These boxes should remain unchecked until the pull request has been -->
<!-- approved. -->

- [ ] Revert dependencies to default branches.
- [ ] Finalize version.

## ✅ Post-merge checklist ##

<!-- Remove any of the following that do not apply. -->

- [ ] Create a release.

chrtorres committed:
Created SaveSearchModal; testing in inventory

chrtorres committed:
Merge branch '543-refactor-of-saved-search-accordion' into save-search-modal

chrtorres committed:
Remove unused import

chrtorres committed:
Troubleshooting modal functions

chrtorres committed:
Troubeshooting API requests Save Search Modal

chrtorres committed:
Merge remote-tracking branch 'origin' into save-search-modal

hawkishpolicy committed:
Save Search modal now saves user defined Search

- moved props, state, and api call from Dashboard to new Saved Search Modal.
- still need to clean up code.
- still need to round out create vuln functionality.

hawkishpolicy committed:
Merge branch 'save-search-modal' of https://github.com/cisagov/XFD into save-search-modal

hawkishpolicy committed:
Restored SaveSearchModal and Stack to Inventory

hawkishpolicy committed:
Merge remote-tracking branch 'origin/develop' into save-search-modal

hawkishpolicy committed:
Modified Save Search button

- Stack is now spaced between.
- Added Save end icon to button.
- Changed button variant to contained for better visibility.

chrtorres committed:
Implemented form validation on save search modal

chrtorres committed:
Added form validation for duplicate entries

chrtorres committed:
Update saved search fixed

chrtorres committed:
Code cleanup; remove unused imports

chrtorres committed:
Code cleanup; remove unused variable

chrtorres committed:
Code cleanup; remove unused useRef

jayjaybunce commented on frontend/src/components/SaveSearchModal/SaveSearchModal.tsx:
```diff
@@ -0,0 +1,254 @@
+      } else {
+        await apiPost('/saved-searches/', body);
+      }
+      history.push('/inventory');
```
I imagine you're reloading the page to ensure we don't have stale saved-searchs in the UI, could potentially change this behavior.
I see that we're using a context for the saved-search data, I think you could pass the fetchSearches into the context so that we could refetch them programmatically instead of reloading the view

chrtorres replied:
I agree. I was keeping the functionality as it was, but refactoring the way this refreshes is a good idea. It's been resolved

jayjaybunce commented on frontend/src/components/SaveSearchModal/SaveSearchModal.tsx:
```diff
@@ -0,0 +1,254 @@
+
+  const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
+    e.preventDefault();
+    console.log(props);
```
💥 💥

chrtorres replied:
Good Catch! This has been resolved.

jayjaybunce reviewed (commented):
Looks and feels great, one small comment on the hard reload of the view.

chrtorres committed:
Removed window.reload with use of state instead; removed unnecessary console logs

hawkishpolicy committed:
Refactored Save Search Accordion

- Removed use of local storage.
- No need as search term, filters, and sort are in context and stored in state.
- Saved Searches are in context and have their own state.
- Created boolean based in state to determine if radio button is checked.

hawkishpolicy committed:
Edited saved-search type for testing

- Commented out lines regarding vuln template and create vuln.

hawkishpolicy committed:
Commented out type def for saved searches at api

hawkishpolicy committed:
Updated saved-searches.test.ts and saved-search.ts

- needed for testing the changes made to the saved-searches.ts file

hawkishpolicy committed:
More test edits

chrtorres committed:
Updated saved-search snapshots; commented out vulnerability fields for saved-search in models

hawkishpolicy committed:
Created new state in SavedSearchContext

- activeSearch is now stored in SavedSearchContext.
- selectedSearch is now stored in SavedSearchContext.
- activeSearchId is now stored in SavedSearchContext.
- activeSearchId is determined in DrawerInterior and used to determine activeSearch and selectedSearch.
- those are then passed to SaveSearchModal.

chrtorres committed:
Fixed update save search; removed search from Inventory; removed existingSearch from SortBar

hawkishpolicy committed:
Cleaned up Saved Search Context and Modal

- Removed unused state from SavedSearchContext and SavedSearchContextProvider.
- Added boolean check to allow user to resuse a name for a saved search if it is the same search.
- Added disbale props to the update search dialogue.

hawkishpolicy committed:
Edits to save src modal, drawer int, & inventory

- Saved Search list is now sorted ascending by name.
- Removed unused imports from SaveSearchModal and Inventory.
- Changed names of dialogs to be more descriptive.
- update handler is now handleUpdate.

chrtorres committed:
Fixed save search/update search placeholder text

chrtorres committed:
Fixed new save search not clearing placeholder text

ameliav commented on frontend/src/components/DrawerInterior.tsx:
```diff
@@ -124,13 +130,16 @@ export const DrawerInterior: React.FC<Props> = (props) => {
+    a.name.localeCompare(b.name)
+  );
+
+  console.log(ascendingSavedSearches);
```
Can the console log be removed?

chrtorres replied:
Log has been removed

jayjaybunce reviewed (approved):
Tested locally, LGTM

ameliav reviewed (commented):
Overall the code and UI look a lot cleaner to me. As for the removed functionality "create a vulnerability", we may want to discuss as a team whether we want to keep it commented or completely remove it.

chrtorres committed:
Removed unnecessary log

cduhn17 reviewed (approved):
As mentioned in the review. Lets get some comments regarding commented out code i.e. the plan for the deprecation of the commented code.

nickviola reviewed (approved):
👍 LGTM!

chrtorres committed:
documented deprecation of create vulnerability function

chrtorres commented:
> As mentioned in the review. Lets get some comments regarding commented out code i.e. the plan for the deprecation of the commented code.

Comments have been added referencing the deprecated code and plans to remove in future versions.

ameliav reviewed (approved):
LGTM with the added reference to the comments!

chrtorres committed:
Added onKeyDown for SaveSearchModal; address 508 compliance requirements

chrtorres committed:
Fixed onKeyDown event not working; Updated onKeyDown to submit form

chrtorres committed:
Added aria labels to forms and buttons within SaveSearchModal

chrtorres committed:
Removed deprecated code

chrtorres committed:
Removed commented code
